File: templates/harness-bench-eval/tasks/L2-dds/L1-PY-02_hello_subscriber/reference/publisher.py
#!/usr/bin/env python3
"""Reference DDS Publisher for HelloWorld topic.

This publisher is used to test AI-generated subscribers.
Uses @idl.struct for type definition (modern API that models tend to use).
"""

import argparse
import time
import sys

import rti.connextdds as dds
import rti.idl as idl


@idl.struct(member_annotations={'message': [idl.bound(256)]})
class HelloWorld:
    message: str = ""
    count: int = 0


def main():
    parser = argparse.ArgumentParser(description="HelloWorld DDS Publisher")
    parser.add_argument("--count", "-c", type=int, default=10, help="Number of samples")
    parser.add_argument("--interval", "-i", type=float, default=0.5, help="Interval between samples")
    parser.add_argument("--domain", "-d", type=int, default=0, help="DDS domain ID")
    args = parser.parse_args()
    
    # Create participant
    participant = dds.DomainParticipant(args.domain)
    
    # Create topic using IDL-defined type
    topic = dds.Topic(participant, "HelloWorld", HelloWorld)
    
    # Create publisher and writer with RELIABLE + TRANSIENT_LOCAL QoS
    publisher = dds.Publisher(participant)
    
    writer_qos = dds.DataWriterQos()
    writer_qos.reliability.kind = dds.ReliabilityKind.RELIABLE
    writer_qos.durability.kind = dds.DurabilityKind.TRANSIENT_LOCAL
    writer_qos.history.kind = dds.HistoryKind.KEEP_ALL
    
    writer = dds.DataWriter(publisher, topic, writer_qos)
    
    # Wait for subscriber discovery
    time.sleep(2.0)
    
    # Publish samples
    for i in range(1, args.count + 1):
        sample = HelloWorld(message="Hello, World!", count=i)
        
        writer.write(sample)
        print(f"Published: message='Hello, World!', count={i}", file=sys.stderr)
        
        if i < args.count:
            time.sleep(args.interval)
    
    # Allow time for reliable delivery
    time.sleep(2.0)
    
    print(f"Published {args.count} samples", file=sys.stderr)


if __name__ == "__main__":
    main()

File: templates/harness-bench-eval/tasks/L2-dds/LN-CPP-01_native_cpp_publisher/reference/publisher.cxx
#include <iostream>
#include <thread>
#include <chrono>
#include <dds/dds.hpp>
#include "HelloWorld.hpp"

int main(int argc, char* argv[]) {
    int count = 10;
    int domain_id = 0;
    
    // Parse command line arguments
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if ((arg == "--count" || arg == "-c") && i + 1 < argc) {
            count = std::stoi(argv[++i]);
        } else if ((arg == "--domain" || arg == "-d") && i + 1 < argc) {
            domain_id = std::stoi(argv[++i]);
        }
    }
    
    try {
        // Create DomainParticipant
        dds::domain::DomainParticipant participant(domain_id);
        
        // Create Topic with HelloWorld type
        dds::topic::Topic<HelloWorld> topic(participant, "HelloWorld");
        
        // Create Publisher
        dds::pub::Publisher publisher(participant);
        
        // Configure QoS for interoperability
        dds::pub::qos::DataWriterQos writer_qos;
        writer_qos << dds::core::policy::Reliability::Reliable()
                   << dds::core::policy::Durability::TransientLocal()
                   << dds::core::policy::History::KeepAll();
        
        // Create DataWriter
        dds::pub::DataWriter<HelloWorld> writer(publisher, topic, writer_qos);
        
        // Wait for subscriber discovery
        std::cerr << "Waiting for discovery..." << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(2));
        
        // Publish samples
        for (int i = 1; i <= count; i++) {
            HelloWorld sample;
            sample.message("Hello, World!");
            sample.count(i);
            
            writer.write(sample);
            std::cerr << "Published: message=\"Hello, World!\", count=" << i << std::endl;
            
            if (i < count) {
                std::this_thread::sleep_for(std::chrono::milliseconds(500));
            }
        }
        
        // Wait for reliable delivery
        std::cerr << "Waiting for delivery..." << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(2));
        
        std::cerr << "Done. Published " << count << " samples." << std::endl;
        
    } catch (const std::exception& ex) {
        std::cerr << "Error: " << ex.what() << std::endl;
        return 1;
    }
    
    return 0;
}
